src/Nerdbank.MessagePack/ByValueEqualityComparer`2.cs

```diff
@@ -0,0 +1,85 @@
+﻿// Copyright (c) Andrew Arnott. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using Nerdbank.MessagePack.SecureHash;
+
+namespace Nerdbank.MessagePack;
+
+/// <summary>
+/// Provides deep by-value implementations of <see cref="IEqualityComparer{T}"/> for arbitrary data types.
+/// </summary>
+/// <typeparam name="T">The data type to be hashed or compared by value.</typeparam>
+/// <typeparam name="TProvider">The provider of the type shape.</typeparam>
+/// <remarks>
+/// <para>
+/// The deep walking of the object graph for deep by-value equality and hashing is based on the same
+/// <see cref="GenerateShapeAttribute"/> that is used to generate MessagePack serializers.
+/// The implementation therefore considers all the same properties for equality and hashing that would
+/// be included in a serialized copy.
+/// </para>
+/// <para>
+/// This implementation is not suitable for all types. Specifically, it is not suitable for types that
+/// have multiple memory representations that are considered equal.
```

Comment:
Does this mean that the comparer might not honor custom `IEquatable<T>` implementations?

Reply:
That's right. Implementing `IEquatable<T>` alone is something some types may already do, but honoring that when the type hasn't provided an equivalent (and/or secure) GetHashCode implementation can split the behavior and violate the invariant, where two distinct value may be seen as equal, yet produce unique hash codes.
There also isn't a guarantee that `IEquatable<T>` was implemented with 'deep by value' semantics.
So I thought it best to define a new interface that guarantees that all these elements are present.